src/libpmemobj/backend_persistent.c

```diff
@@ -0,0 +1,464 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * backend_persistent.c -- implementation of persistent backend
+ */
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <libpmem.h>
+#include <string.h>
+#include "bucket.h"
+#include "arena.h"
+#include "backend.h"
+#include "pool.h"
+#include "util.h"
+#include "out.h"
+#include "backend_persistent.h"
+
+static struct bucket_backend_operations persistent_bucket_ops = {
+
+};
+
+static struct arena_backend_operations persistent_arena_ops = {
+	.set_alloc_ptr = persistent_set_alloc_ptr
+};
+
+static struct pool_backend_operations persistent_pool_ops = {
+
+};
+
+/*
+ * verify_header -- (internal) check is the header is consistent
+ */
+static bool
+verify_header(struct backend_pool_header *h)
+{
+	if (util_checksum(h, sizeof (*h), &h->checksum, 0) != 1) {
+		return false;
+	}
+
+	if (memcmp(h->signature, POOL_SIGNATURE, POOL_SIGNATURE_LEN) != 0) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * copy_header -- (internal) create a copy of a header
+ */
+static void
+copy_header(struct backend_persistent *b, struct backend_pool_header *left,
+	struct backend_pool_header *right)
+{
+	memcpy(left, right, sizeof (*left));
+	b->persist(left, sizeof (*left));
+}
+
+/*
+ * recover_primary_header -- (internal) check backups for a valid header copy
+ */
+static bool
+recover_primary_header(struct backend_persistent *b)
+{
+	for (int i = 0; i < b->max_zone; ++i) {
+		if (verify_header(&b->pool->zone[i].backup_header)) {
+			copy_header(b, &b->pool->primary_header,
+				&b->pool->zone[i].backup_header);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/*
+ * zero_info_slots -- (internal) zerofill all info slot structures
+ */
+static void
+zero_info_slots(struct backend_persistent *b)
+{
+	memset(b->pool->info_slot, 0, sizeof (b->pool->info_slot));
+	b->persist(b->pool->info_slot, sizeof (b->pool->info_slot));
+}
+
+/*
+ * write_primary_pool_header -- (internal) create a fresh pool header
+ */
+static void
+write_primary_pool_header(struct backend_persistent *b)
+{
+	struct backend_pool_header *hdrp = &b->pool->primary_header;
+	memcpy(hdrp->signature, POOL_SIGNATURE, POOL_SIGNATURE_LEN);
+	hdrp->flags = 0;
+	hdrp->state = POOL_STATE_CLOSED;
+	hdrp->major = PERSISTENT_BACKEND_MAJOR;
+	hdrp->minor = PERSISTENT_BACKEND_MINOR;
+	hdrp->size = b->pool_size;
+	hdrp->chunk_size = CHUNKSIZE;
+	hdrp->chunks_per_zone = MAX_CHUNK;
+	memset(hdrp->reserved, 0, sizeof (hdrp->reserved));
+	hdrp->checksum = 0;
+	b->persist(hdrp, sizeof (*hdrp));
+	util_checksum(hdrp, sizeof (*hdrp), &hdrp->checksum, 1);
```

Comment:
What happens if we crash after the util_checksum but before we have the chance to really persist the data? The pool header is invalid and we do what?